enhancements/machine-api/azure-ultra-disks.md

```diff
@@ -0,0 +1,243 @@
+
+### User Stories
+
+**Story 1**
+As a developer, I'd like to have the ability to use Azure Ultra Disks as Data Disks for my Machines.
+
+**Story 2**
+As a developer, I'd like to specify some basic specs (Disk Size, Logical Unit Number (LUN), Caching Type) for creating Azure Ultra Disks as Data Disks for my Machines.
+
+**Story 3**
+As a developer, I'd like to have the ability to create Persistent Volume Claims (PVCs) which can dynamically bind to a StorageClass backed by Azure Ultra Disk and to mount them successfully to Pods for my workloads needs.
+
+### API Extensions
+
+#### Extension for Ultra Disks as Data Disks
+
+The API design is taken from the upstream Cluster API Provider for Azure (CAPZ), [here](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/fee74172efc0db07a244ceb870be759db56e0f83/api/v1beta1/azuremachine_types.go#L77-L79) and [here](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/fee74172efc0db07a244ceb870be759db56e0f83/api/v1beta1/types.go#L465-L483).
+The upstream design is followed as closely as possible to ensure smoother future migrations between MAPI and CAPI.
+
+A new field will be added to the `AzureMachineProviderSpec` struct `DataDisks`. The interface will be a slice of `DataDisk` structs allowing multiple Data Disks and various fields specific to them to be further specified.
+
+A `DataDisk` has one Composite type field, `ManagedDisk` of type `*ManagedDiskParameters`, which is a type already defined in the Machine API spec, used for the `ManagedDisk` field in the `OSDisk` type struct, another `AzureMachineProviderSpec` field.
+
+
+```go
+type AzureMachineProviderSpec struct {
+  // Existing fields will not be modified
+  ...
+
+  // DataDisk specifies the parameters that are used to add one or more data disks to the machine
+  // +optional
+  DataDisks []DataDisk `json:"dataDisks,omitempty"`
+}
+
+// DataDisk specifies the parameters that are used to add one or more data disks to the machine.
+type DataDisk struct {
+  // NameSuffix is the suffix to be appended to the machine name to generate the disk name.
+  // Each disk name will be in format <machineName>_<nameSuffix>.
+  NameSuffix string `json:"nameSuffix"`
+  // DiskSizeGB is the size in GB to assign to the data disk.
+  DiskSizeGB int32 `json:"diskSizeGB"`
+  // ManagedDisk specifies the Managed Disk parameters for the data disk.
+  // +optional
+  ManagedDisk *ManagedDiskParameters `json:"managedDisk,omitempty"`
+  // Lun Specifies the logical unit number of the data disk. This value is used to identify data disks within the VM and therefore must be unique for each data disk attached to a VM.
+  // The value must be between 0 and 63.
```

Comment:
Can we add kubebuilder tags that specify this

Reply:
does this field have to be a pointer?

Reply:
> does this field have to be a pointer?

Good point, it is a pointer [upstream](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/b5c7a99f435ff907fd9bab3e993dc0db439fe329/api/v1beta1/types.go#L474), but if we look at how we use it downstream in other places like in [`AzureMachineProviderSpec`'s `OSDisk`](https://github.com/openshift/api/blob/d74727069f6fb9193d6ba50ed7e3a5edb2477bec/machine/v1beta1/types_azureprovider.go#L203) that's not defined as a pointer. What convention should we follow?

Reply:
`0` is a valid value for the LUN, as is unset, so we need a way to distinguish between them, this has to be a pointer to do that

Reply:
Indeed.